src/Internal/Payment/Payment.php

```diff
@@ -0,0 +1,137 @@
+<?php
+/**
+ * Class Payment
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Internal\Payment;
+
+use WC_Order;
+use WCPay\Internal\Payment\PaymentMethod\PaymentMethodInterface;
+
+/**
+ * Payment object.
+ *
+ * This is a data object, used as context and storage for payment states.
+ */
+class Payment {
+	/**
+	 * Order requiring/having payment.
+	 *
+	 * @var WC_Order
+	 */
+	private $order;
+
+	/**
+	 * Holds all data, related to the payment.
+	 *
+	 * @var array
+	 */
+	private $data = [];
+
+	/**
+	 * Payment constructor.
+	 *
+	 * @param WC_Order $order Order requiring payment.
+	 */
+	public function __construct( WC_Order $order ) {
+		$this->order = $order;
+	}
+
+	/**
+	 * Returns the payment's order.
+	 *
+	 * @return WC_Order
+	 */
+	public function get_order() {
+		return $this->order;
+	}
+
+	/**
+	 * Stores data within the payment.
+	 *
+	 * @param string $key   Key of the property to store.
+	 * @param mixed  $value Value to store.
+	 */
+	private function store( string $key, $value ) {
+		$this->data[ $key ] = $value;
+	}
+
+	/**
+	 * Reads from the internal data array.
+	 *
+	 * @param string $key Key to read.
+	 * @return mixed|null
+	 */
+	private function read( string $key ) {
+		return $this->data[ $key ] ?? null;
+	}
+
+	/**
+	 * Adds a flag to the payment.
+	 *
+	 * Flags are used to control small nuances in the payment
+	 * processing, rather than controlling the flow.
+	 *
+	 * @param Flag $flag The flag to add.
+	 */
+	public function add_flag( Flag $flag ) {
+		$flags = $this->read( 'flags' ) ?? [];
+
+		if ( ! in_array( $flag, $flags, true ) ) {
+			$flags[] = $flag;
+			$this->store( 'flags', $flags );
```

Comment:
> I'm still wondering why we need flags.

A couple of reasons:

1. Flags represent minor deviations in the payment process, which do not warrant having separate code flows for them.
2. If those are not fags, but rather simple parameters, they would need to be passed around quite a bit until they reach the method/service, which actually cares about them. Also, it might make calling methods like `->process( $manual_capture, true, false, false, true )`.
3. By defining some of those flags in advance, and storing them within the context, we'd be able to reuse them later. For example, checking whether to save the payment method within the post-3DS-authentication flow.

> Also, what do you think about having an array of booleans?

For me, a flag is either present or not. If it is not present, keeping it in the array doesn't make too much sense. Maybe using `Base_Contant` here was useless though, I'll just switch to a plain class with constants.